### PR TITLE
Add support for `name`, `pkgs` and `diff_attr` parameters to `zypperpkg.upgrade()`/`yumpkg.upgrade()`

### LINKED ISSUE #62030: [BUG] zypperpkg.upgrade does not support name or pkgs parameters
**Description**

`pkg.upgrade name=foo` works with `yumpkg`, but not with `zypperpkg`. Different `pkg` implementations should try to support the same arguments where it makes sense. Passing `name` or `pkgs` to limit the number of packages that will be upgraded makes sense to a Salt user.

**Setup**

One minion with `yum`, e.g. AlmaLinux 8 and another with `zypper` (e.g. openSUSE Leap 15.3). Salt 3004 on the master and the minions.
 
**Steps to Reproduce the behavior**:
`salt '*' pkg.upgrade name=foo` and `salt '*' pkg.upgrade pkgs='["foo", "bar"]'`:
In both cases, the minion running `zypper` ignores the parameter that restricts the packages to be upgraded and upgrades all packages.

**Expected behavior**:

`zypper` only update the packages that are specified, the same way that `yum` does it. If dependencies need to be updated along with the specified packages, they will also be updated. All updated packages are reported in the job return dict and rendered on the CLI.

**Versions Report**
<details><summary>salt --versions-report</summary>
The exact version does not matter, `zypperpkg.upgrade` does not know the parameter in any of them. I reproduced this on a setup using

```yaml
Salt Version:
          Salt: 3002.2
 
Dependency Versions:
          cffi: 1.13.2
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: 0.28.4
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.17
      pycrypto: 3.9.0
  pycryptodome: Not Installed
        pygit2: 0.28.2
        Python: 3.6.13 (default, Mar 10 2021, 18:30:35) [GCC]
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.3
 
System Versions:
          dist: sles 15.3 2021.05
        locale: UTF-8
       machine: x86_64
       release: 5.3.18-59.27-default
        system: Linux
       version: SLES 15.3 2021.05

```
</details>

**Additional context**

### LINKED ISSUE #62031: [FEATURE REQUEST] Add `diff_attr` to `pkg.upgrade()` (At least for Yum, Zypper, Apt)
**Is your feature request related to a problem? Please describe.**
`pkg.upgrade()` only provides limited information of packages that changed, while `pkg.install()` can, optionally, provide more information. The additional information can also be gathered by `pkg.list_pkgs(attr=X)` where `X` is a list of attributes or the string "all", but that requires an additional function call.

**Describe the solution you'd like**
Just like `pkg.install(..., diff_attr=X)`, `pkg.upgrade(..., diff_attr=X)` returns the specified attributes in addition to the name and the version.

**Describe alternatives you've considered**
Since this addition is very small and already exists in `pkg.install()`, I haven't considered alternatives to adding `diff_attr` to `pkg.upgrade()`. I'm not aware of any downside except for having another parameter in `pkg.upgrade()`.

### LINKED ISSUE #62032: [BUG] `pkg.list_pkgs(attr="all`)` (yum/zypper) calsl `pkg_resource.format_pkg_list()` wrongly
**Description**
`pkg.list_pkgs(attr="all")` turns `"all"` into `["all"]` before passing it to `pkg_resource.format_pkg_list()`. The latter function expects either a list of attributes or the string "all". Upon encountering `["all"]`, it parses it as a list of attributes to include in its return value. As a result, only the default "version" and "arch" attributes are returned.

**Setup**

One minion with `yum`, e.g. AlmaLinux 8 and another with `zypper` (e.g. openSUSE Leap 15.3). Salt 3004 on the master and the minions.
 
**Steps to Reproduce the behavior**:
```
# salt '*'  pkg.list_pkgs attr=all
[...]
    zlib:
        |_
          ----------
          arch:
              x86_64
          version:
              1.2.11
```

**Expected behavior**:
```
# salt '*'  pkg.list_pkgs attr=all
[...]
    zypper:
        |_
          ----------
          arch:
              x86_64
          epoch:
              None
          install_date:
              2022-02-22T18:22:09Z
          install_date_time_t:
              1645554129
          release:
              24.1
          version:
              1.14.51

```

**Versions Report**
<details><summary>salt --versions-report</summary>
The exact version does not matter, `zypperpkg.upgrade` does not know the parameter in any of them. I reproduced this on a setup using

```yaml
Salt Version:
          Salt: 3002.2
 
Dependency Versions:
          cffi: 1.13.2
      cherrypy: unknown
      dateutil: 2.8.1
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 2.10.1
       libgit2: 0.28.4
      M2Crypto: 0.35.2
          Mako: Not Installed
       msgpack: 0.5.6
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: 2.17
      pycrypto: 3.9.0
  pycryptodome: Not Installed
        pygit2: 0.28.2
        Python: 3.6.13 (default, Mar 10 2021, 18:30:35) [GCC]
  python-gnupg: Not Installed
        PyYAML: 5.4.1
         PyZMQ: 17.0.0
         smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.2.3
 
System Versions:
          dist: sles 15.3 2021.05
        locale: UTF-8
       machine: x86_64
       release: 5.3.18-59.27-default
        system: Linux
       version: SLES 15.3 2021.05

```
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by agraul**
### What does this PR do?
Add `diff_attr` to `yumpkg.upgrade()` and add it, together with `name` and `pkgs`, to `zypperpkg.upgrade()`.

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/62030
Fixes: https://github.com/saltstack/salt/issues/62031
Fixes: https://github.com/saltstack/salt/issues/62032

### Previous Behavior
`name` and `pkgs` parameters are ignored in `zypperpkg.upgrade()`. The `diff_attr` parameter is ignored by both `yumpkg.upgrade()` and `zypperpkg.upgrade()`

### New Behavior
`name` and `pkgs` parameters are used in `zypperpkg.upgrade` to only update/upgrade the specified package(s). `diff_attr` is used by `zypperpkg.upgrade()` and `yumpkg.upgrade()` to output additional attributes.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by SchoolGuy — DISMISSED** [review]
I have debugged this together with @agraul. The PR is LGTM for me.

**Comment by Ch3LL** [comment]
looks like there is a test failure that might be related to this change

**Comment by agraul** [comment]
> looks like there is a test failure that might be related to this change

After rebasing the PR on top of the current master commit, I don't see the test failure locally anymore. Let's see if jenkins' test executions agree.

File: salt/modules/yumpkg.py [hunk 1]
```diff
@@ -1834,6 +1834,7 @@ def upgrade(
     normalize=True,
     minimal=False,
     obsoletes=True,
+    diff_attr=None,
```
Comment (Ch3LL) [inline salt/modules/yumpkg.py:1837]: can you add documentation below for this new argument

**Review by Ch3LL — CHANGES_REQUESTED** [review]

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**